File: 0743. Network Delay Time/Solution.py
import heapq
from math import inf
from typing import List


class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        graph = [[] for _ in range(n)]
        for u, v, d in times:
            graph[u - 1].append((v - 1, d))

        dis = [inf] * n
        dis[k - 1] = 0
        pq = [(0, k - 1)]
        while pq:
            du, u = heapq.heappop(pq)
            if du > dis[u]:
                continue
            for v, d in graph[u]:
                new_dis = du + d
                if new_dis < dis[v]:
                    dis[v] = new_dis
                    heapq.heappush(pq, (new_dis, v))
        mx = max(dis)
        return mx if mx < inf else -1

File: 0743. Network Delay Time/Solution.cpp
#include <algorithm>
#include <queue>
#include <ranges>
#include <vector>
using namespace std;

class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k)
    {
        vector<vector<pair<int, int>>> graph(n);
        for (const vector<int>& t : times)
            graph[t[0] - 1].emplace_back(t[1] - 1, t[2]);

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
        pq.emplace(0, k - 1);
        vector<int> dis(n, INT_MAX);
        dis[k - 1] = 0;

        while (!pq.empty()) {
            auto [du, u] = pq.top();
            pq.pop();

            if (du > dis[u])
                continue;

            for (const auto& [v, d] : graph[u]) {
                int new_dis = du + d;
                if (new_dis < dis[v]) {
                    dis[v] = new_dis;
                    pq.emplace(new_dis, v);
                }
            }
        }

        int mx = ranges::max(dis);
        return mx == INT_MAX ? -1 : mx;
    }
};
